sound/soc/sof/intel/telemetry.c

```diff
@@ -0,0 +1,168 @@
+			telemetry_data->separator_1);
+		return false;
+	}
+
+	if (telemetry_data->separator_2 != SOF_TELEMETRY_SEPARATOR_2) {
+		dev_err(sdev->dev, "error: separator_2 %#x is not matched",
+			telemetry_data->separator_2);
+		return false;
+	}
+
+	if (telemetry_data->separator_3 != SOF_TELEMETRY_SEPARATOR_3) {
+		dev_err(sdev->dev, "error: separator_3 %#x is not matched",
+			telemetry_data->separator_3);
+		return false;
+	}
+
+	if (telemetry_data->separator_4 != SOF_TELEMETRY_SEPARATOR_4) {
+		dev_err(sdev->dev, "error: separator_4 %#x is not matched",
+			telemetry_data->separator_4);
+		return false;
+	}
+
+	if (telemetry_data->separator_5 != SOF_TELEMETRY_SEPARATOR_5) {
+		dev_err(sdev->dev, "error: separator_5 %#x is not matched",
+			telemetry_data->separator_5);
+		return false;
+	}
+
+	if (telemetry_data->separator_6 != SOF_TELEMETRY_SEPARATOR_6) {
+		dev_err(sdev->dev, "error: separator_6 %#x is not matched",
+			telemetry_data->separator_6);
+		return false;
+	}
+
+	if (telemetry_data->separator_7 != SOF_TELEMETRY_SEPARATOR_7) {
+		dev_err(sdev->dev, "error: separator_7 %#x is not matched",
+			telemetry_data->separator_7);
+		return false;
+	}
+
+	if (telemetry_data->separator_8 != SOF_TELEMETRY_SEPARATOR_8) {
+		dev_err(sdev->dev, "error: separator_8 %#x is not matched",
+			telemetry_data->separator_8);
+		return false;
+	}
+
+	if (telemetry_data->separator_9 != SOF_TELEMETRY_SEPARATOR_9) {
+		dev_err(sdev->dev, "error: separator_9 %#x is not matched",
+			telemetry_data->separator_9);
+		return false;
+	}
+
+	if (telemetry_data->separator_10 != SOF_TELEMETRY_SEPARATOR_10) {
+		dev_err(sdev->dev, "error: separator_10 %#x is not matched",
+			telemetry_data->separator_10);
+		return false;
+	}
+
+	if (telemetry_data->separator_11 != SOF_TELEMETRY_SEPARATOR_11) {
+		dev_err(sdev->dev, "error: separator_11 %#x is not matched",
+			telemetry_data->separator_11);
+		return false;
+	}
+
+	if (telemetry_data->separator_exception != SOF_TELEMETRY_SEPARATOR_EXP) {
+		dev_err(sdev->dev, "error: separator_exception %#x is not matched",
+			telemetry_data->separator_exception);
+		return false;
+	}
+
+	return true;
+}
+
+void sof_ipc4_intel_dump_telemetry_state(struct snd_sof_dev *sdev, u32 flags)
+{
+	char *level = (flags & SOF_DBG_DUMP_OPTIONAL) ? KERN_DEBUG : KERN_ERR;
+	struct telemetry_wnd_data *telemetry_data;
+	struct sof_ipc_dsp_oops_xtensa *xoops;
+	u32 slot_offset;
+	u32 core;
+
+	telemetry_data = kmalloc(sizeof(*telemetry_data), GFP_KERNEL);
+	if (!telemetry_data)
+		return;
+
+	xoops = kzalloc(sizeof(*xoops) + SOF_IPC4_FW_AREGS_COUNT * sizeof(int), GFP_KERNEL);
+	if (!xoops)
+		goto free;
+
+	slot_offset = find_telemetry_slots(sdev);
+	if (!slot_offset)
+		goto free;
+
+	sof_mailbox_read(sdev, slot_offset, telemetry_data, sizeof(*telemetry_data));
+
+	if (!validate_telemetry_data(sdev, telemetry_data))
+		goto free;
+
+	core = (flags & SOF_DBG_DUMP_CORE_MASK) >> SOF_DBG_DUMP_CORE_SHIFT;
```

Comment:
isn't there a macro for this already?

FIELD_GET() is what you want!

Reply:
thanks, I will try